paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/orphan/SparkOrphanFilesClean.scala

```diff
@@ -0,0 +1,223 @@
+ * limitations under the License.
+ */
+
+package org.apache.paimon.spark.orphan
+
+import org.apache.paimon.Snapshot
+import org.apache.paimon.catalog.{Catalog, Identifier}
+import org.apache.paimon.fs.Path
+import org.apache.paimon.manifest.{ManifestEntry, ManifestFile, ManifestFileMeta}
+import org.apache.paimon.operation.OrphanFilesClean
+import org.apache.paimon.operation.OrphanFilesClean.retryReadingFiles
+import org.apache.paimon.table.FileStoreTable
+import org.apache.paimon.utils.SerializableConsumer
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{Dataset, SparkSession}
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.functions.sum
+
+import java.util
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicLong
+import java.util.function.Consumer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+case class SparkOrphanFilesClean(
+    specifiedTable: FileStoreTable,
+    specifiedOlderThanMillis: Long,
+    specifiedFileCleaner: SerializableConsumer[Path],
+    parallelism: Int,
+    @transient spark: SparkSession)
+  extends OrphanFilesClean(specifiedTable, specifiedOlderThanMillis, specifiedFileCleaner)
+  with SQLConfHelper
+  with Logging {
+
+  def doOrphanClean(): Dataset[Long] = {
+    import spark.implicits._
+
+    val branches = validBranches()
+    val deletedInLocal = new AtomicLong(0)
+    // snapshot and changelog files are the root of everything, so they are handled specially
+    // here, and subsequently, we will not count their orphan files.
+    cleanSnapshotDir(branches, (_: Path) => deletedInLocal.incrementAndGet)
+
+    val maxBranchParallelism = Math.min(branches.size(), parallelism)
+    // find snapshots using branch and find manifests(manifest, index, statistics) using snapshot
+    val usedManifestFiles = spark.sparkContext
+      .parallelize(branches.asScala, maxBranchParallelism)
+      .mapPartitions(_.flatMap {
+        branch => safelyGetAllSnapshots(branch).asScala.map(snapshot => (branch, snapshot.toJson))
+      })
+      .repartition(parallelism)
+      .flatMap {
+        case (branch, snapshotJson) =>
+          var isManifestMetaFile: Boolean = false
+          val usedFileBuffer = new ArrayBuffer[BranchAndManifestFile]()
+          val usedFileConsumer = new Consumer[String] {
+            override def accept(file: String): Unit = {
+              usedFileBuffer.append(BranchAndManifestFile(branch, file, isManifestMetaFile))
+              isManifestMetaFile = false
+            }
+          }
+          val manifestConsumer = new Consumer[ManifestFileMeta] {
+            override def accept(t: ManifestFileMeta): Unit = {
+              isManifestMetaFile = true
+            }
+          }
+          val snapshot = Snapshot.fromJson(snapshotJson)
+          collectWithoutDataFile(branch, snapshot, usedFileConsumer, manifestConsumer)
+          usedFileBuffer
+      }
+      .toDS()
```

Comment:
Should we cache this DS to avoid computing twice?

Reply:
addressed